File: binding-go/runtime/src/main/go/etch/DefaultValueFactory.go
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements. See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership. The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License. You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package etch

import "fmt"
import "reflect"

const (
	ETCH_RUNTIME_EXCEPTION_TYPE_NAME string = "_Etch_RuntimeException"
	ETCH_LIST_TYPE_NAME              string = "_Etch_List"
	ETCH_MAP_TYPE_NAME               string = "_Etch_Map"
	ETCH_SET_TYPE_NAME               string = "_Etch_Set"
	ETCH_DATETIME_TYPE_NAME          string = "_Etch_Datetime"
	ETCH_AUTH_EXCEPTION_TYPE_NAME    string = "_Etch_AuthException"
	ETCH_EXCEPTION_MESSAGE_NAME      string = "_exception"
	MSG_FIELD_NAME                   string = "msg"
	MESSAGE_ID_FIELD_NAME            string = "_messageId"
	IN_REPLY_TO_FIELD_NAME           string = "_inReplyTo"
	RESULT_FIELD_NAME                string = "result"
)

type DefaultValueFactory struct {
	Types                      *TypeMap
	Class2type                 *map[reflect.Type]*Type
	fieldMap                   map[string]*Field
	_mf_msg                    *Field
	_mf__messageId             *Field
	_mf__inReplyTo             *Field
	_mf_result                 *Field
	_mt__Etch_RuntimeException *Type
	_mt__Etch_List             *Type
	_mt__Etch_Map              *Type
	_mt__Etch_Set              *Type
	_mt__Etch_Datetime         *Type
	_mt__Etch_AuthException    *Type
	_mt__exception             *Type
}

func NewDefaultValueFactory() *DefaultValueFactory {
	ret := new(DefaultValueFactory)
	ret.Types = NewTypeMap()
	ret.Class2type = &(map[reflect.Type]*Type{})
	ret.fieldMap = map[string]*Field{}
	ret._mf_msg = NewField(MSG_FIELD_NAME)
	ret._mf__messageId = NewField(MESSAGE_ID_FIELD_NAME)
	ret._mf__inReplyTo = NewField(IN_REPLY_TO_FIELD_NAME)
	ret._mf_result = NewField(RESULT_FIELD_NAME)
	ret._mt__Etch_RuntimeException = ret.Types.GetByName(ETCH_RUNTIME_EXCEPTION_TYPE_NAME)
	ret._mt__Etch_List = ret.Types.GetByName(ETCH_LIST_TYPE_NAME)
	ret._mt__Etch_Map = ret.Types.GetByName(ETCH_MAP_TYPE_NAME)
	ret._mt__Etch_Set = ret.Types.GetByName(ETCH_SET_TYPE_NAME)
	ret._mt__Etch_Datetime = ret.Types.GetByName(ETCH_DATETIME_TYPE_NAME)
	ret._mt__Etch_AuthException = ret.Types.GetByName(ETCH_AUTH_EXCEPTION_TYPE_NAME)
	ret._mt__exception = ret.Types.GetByName(ETCH_EXCEPTION_MESSAGE_NAME)
	return ret
}

func (dvf *DefaultValueFactory) AddField(field *Field) {
	dvf.fieldMap[field.Name] = field
}

func (dvf *DefaultValueFactory) GetFieldByName(name string) *Field {
	return dvf.fieldMap[name]
}

func (dvf *DefaultValueFactory) GetClass2type() *map[reflect.Type]*Type {
	return dvf.Class2type
}


func (dvf *DefaultValueFactory) GetDefaultResponseField() *Field {
	return dvf._mf_result
}

func (dvf *DefaultValueFactory) GetDefaultInReplyToField() *Field {
	return dvf._mf__inReplyTo
}

func (dvf *DefaultValueFactory) GetDefaultMessageIdField() *Field {
	return dvf._mf__messageId
}


func (dvf *DefaultValueFactory) GetTypeById(id int32) *Type {
	ty := dvf.Types.GetById(id)
	return ty
}

func (dvf *DefaultValueFactory) GetTypeByName(name string) *Type {
	ty := dvf.Types.GetByName(name)
	return ty
}

func (dvf *DefaultValueFactory) AddType(ty *Type) {
	Log("DefaultValueFactory: adding type " + fmt.Sprint(ty) + "\n")
	dvf.Types.Add(ty)
}

func ToLong(val interface{}) int64 {
	switch val.(type) {
	case int:
		return int64(val.(int))
	case int8:
		return int64(val.(int8))
	case int16:
		return int64(val.(int16))
	case int32:
		return int64(val.(int32))
	case int64:
		return int64(val.(int64))
	}
	return 0
}

func (dvf *DefaultValueFactory) GetMessageId(theMsg *Message) int64 {
	return ToLong(theMsg.Fields[dvf._mf__messageId])
}

func (dvf *DefaultValueFactory) SetMessageId(msg *Message, msgId int64) {
	Log("DefaultValueFactory SETMESSAGEID " + fmt.Sprint(msgId) + "\n")
	msg.Fields[dvf._mf__messageId] = msgId
}

func (dvf *DefaultValueFactory) GetTypes() []*Type {
	tys := dvf.Types.GetAll()
	newslice := make([]*Type, len(tys))
	copy(newslice, tys)
	return newslice
}

func (dvf *DefaultValueFactory) GetInReplyTo(msg *Message) int64 {
	val := msg.Fields[dvf._mf__inReplyTo]
	Log("DefaultValueFactory GetInReplyTo looking in fields " + fmt.Sprint(msg.Fields) + "/" + fmt.Sprint(&(dvf._mf__inReplyTo)) + "\n")
	if val == nil {
		Log("DefaultValueFactory got nil \n")
		return 0
	}
	Log("DefaultValueFactory GetInReplyTo was " + fmt.Sprint(val) + "\n")
	return ToLong(val)
}

func (dvf *DefaultValueFactory) SetInReplyTo(msg *Message, msgId int64) {
	msg.Fields[dvf._mf__inReplyTo] = msgId
}

func (dvf *DefaultValueFactory) ExportCustomValue(value interface{}) *StructValue {
	ty := reflect.TypeOf(value)
	etchty := (*dvf.Class2type)[ty]
	if etchty.Helper != nil {
		return etchty.Helper.ExportValue(etchty, dvf, value)
	}
	return nil
}

func (dvf *DefaultValueFactory) ImportCustomValue(value *StructValue) interface{} {
	if value.Ty.Helper != nil {
		return value.Ty.Helper.ImportValue(value)
	}
	return nil
}

func (dvf *DefaultValueFactory) GetCustomStructType(obj reflect.Type) *Type {
	return (*dvf.Class2type)[obj]
}


func NewboolArray(length int32) []bool {
	return make([]bool, length)
}
func Newint8Array(length int32) []int8 {
	return make([]int8, length)
}
func Newint16Array(length int32) []int16 {
	return make([]int16, length)
}
func Newint32Array(length int32) []int32 {
	return make([]int32, length)
}
func Newint64Array(length int32) []int64 {
	return make([]int64, length)
}
func Newfloat32Array(length int32) []float32 {
	return make([]float32, length)
}
func Newfloat64Array(length int32) []float64 {
	return make([]float64, length)
}
func NewstringArray(length int32) []interface{} {
	return make([]interface{}, length)
}
func NewInterfaceArray(length int32) []interface{} {
	return make([]interface{}, length)
}

func (dvf *DefaultValueFactory)NewArrayForType(typeCode int8, cTy *Type, length int32) interface{}{
	switch typeCode {
	case BYTE:
		return Newint8Array(length)
	case SHORT:
		return Newint16Array(length)
	case INT:
		return Newint32Array(length)
	case LONG:
		return Newint64Array(length)
	case FLOAT:
		return Newfloat32Array(length)
	case DOUBLE:
		return Newfloat64Array(length)
	case STRING:
		return NewstringArray(length)
	case CUSTOM:
		return cTy.NewArray(length)
	}
	return nil
}

File: binding-csharp/runtime/src/main/csharp/Org.Apache.Etch.Bindings.Csharp/Support/DefaultValueFactory.cs
// $Id$
// 
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements. See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership. The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License. You may obtain a copy of the License at
// 
//   http://www.apache.org/licenses/LICENSE-2.0
// 
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied. See the License for the
// specific language governing permissions and limitations
// under the License.
// 
using System;
using System.Collections;
using System.Collections.Generic;
using System.Text;
using Org.Apache.Etch.Bindings.Csharp.Msg;
using Org.Apache.Etch.Bindings.Csharp.Util;

namespace Org.Apache.Etch.Bindings.Csharp.Support
{

    /// <summary>Default implementation of ValueFactory which provides some dynamic type and field support, as well 
    /// as standard value conversions and import and export.</summary>
    public class DefaultValueFactory : ValueFactory
    {
        private const String ETCH_RUNTIME_EXCEPTION_TYPE_NAME = "_Etch_RuntimeException";

        private const String ETCH_LIST_TYPE_NAME = "_Etch_List";

	    private const String ETCH_MAP_TYPE_NAME = "_Etch_Map";

	    private const String ETCH_SET_TYPE_NAME = "_Etch_Set";

	    private const String ETCH_DATETIME_TYPE_NAME = "_Etch_Datetime";

	    private const String ETCH_AUTH_EXCEPTION_TYPE_NAME = "_Etch_AuthException";

	    private const String ETCH_EXCEPTION_MESSAGE_NAME = "_exception";

	    private const String MSG_FIELD_NAME = "msg";

	    private const String MESSAGE_ID_FIELD_NAME = "_messageId";

	    private const String IN_REPLY_TO_FIELD_NAME = "_inReplyTo";

        private const String RESULT_FIELD_NAME = "result";
     

        public static void Init( TypeMap types, Class2TypeMap class2type )
        {
            RuntimeExceptionSerializer.Init(
             types.Get(ETCH_RUNTIME_EXCEPTION_TYPE_NAME), class2type);

            ListSerializer.Init(
                types.Get(ETCH_LIST_TYPE_NAME), class2type);

            MapSerializer.Init(
                types.Get(ETCH_MAP_TYPE_NAME), class2type);

        /*    SetSerializer.Init(
                types.Get(ETCH_SET_TYPE_NAME), class2type); */

            DateSerializer.Init(
                types.Get(ETCH_DATETIME_TYPE_NAME), class2type);

            AuthExceptionSerializer.Init(
                types.Get(ETCH_AUTH_EXCEPTION_TYPE_NAME), class2type);
         
            XType t3 = types.Get(ETCH_EXCEPTION_MESSAGE_NAME);
            t3.PutValidator(_mf_result, Validator_RuntimeException.Get());
            t3.PutValidator(_mf__messageId, Validator_long.Get(0));
            t3.PutValidator(_mf__inReplyTo, Validator_long.Get(0));
        }

        /// <summary>The msg field of the standard unchecked exception.</summary>
        public readonly static Field _mf_msg = new Field(MSG_FIELD_NAME);

        /// <summary>The well-known _messageId field.</summary>
        public readonly static Field _mf__messageId = new Field(MESSAGE_ID_FIELD_NAME);

        /// <summary>The well-known _inReplyTo field. </summary>
        public readonly static Field _mf__inReplyTo = new Field(IN_REPLY_TO_FIELD_NAME);

        /// <summary> The well-known result field </summary>
        public readonly static Field _mf_result = new Field(RESULT_FIELD_NAME);

        private readonly TypeMap types;

        private readonly TypeMap dynamicTypes = new TypeMap();

	    private readonly Class2TypeMap class2type;
	
	    private readonly XType _mt__Etch_RuntimeException;
	
	    private readonly XType _mt__Etch_AuthException;
	
	    private readonly XType _mt__exception;

        private readonly XType _mt__Etch_List;
	
	    private readonly XType _mt__Etch_Map;
	
	    private readonly XType _mt__Etch_Set;
	
	    private readonly XType _mt__Etch_Datetime;
	

       /// <summary>
        /// Constructs the DefaultValueFactory.
       /// </summary>
       /// <param name="uri"></param>
       /// <param name="types"></param>
       /// <param name="class2type"></param>
        public DefaultValueFactory(String uri, TypeMap types, Class2TypeMap class2type)
        {
            URL u = new URL(uri);
            
            String s = u.GetTerm("DefaultValueFactory.level", "FULL");
            level = (Validator.Level) Enum.Parse(typeof(Validator.Level), s);

            this.types = types;
            this.class2type = class2type;

            _mt__Etch_RuntimeException = types.Get(ETCH_RUNTIME_EXCEPTION_TYPE_NAME);
            _mt__Etch_AuthException = types.Get(ETCH_AUTH_EXCEPTION_TYPE_NAME);
            _mt__exception = types.Get(ETCH_EXCEPTION_MESSAGE_NAME);
            _mt__Etch_List = types.Get(ETCH_LIST_TYPE_NAME);
            _mt__Etch_Map = types.Get(ETCH_MAP_TYPE_NAME);
            _mt__Etch_Set = types.Get(ETCH_SET_TYPE_NAME);
            _mt__Etch_Datetime = types.Get(ETCH_DATETIME_TYPE_NAME);
        }

        public XType get_mt__Etch_RuntimeException()
	    {
		    return _mt__Etch_RuntimeException;
	    }
	
	    public XType get_mt__Etch_AuthException()
	    {
		    return _mt__Etch_AuthException;
	    }
	
	    public  XType get_mt__exception()
	    {
		    return _mt__exception;
	    }

        public XType Get_mt__Etch_List()
	    {
		    return _mt__Etch_List;
	    }
	
	    public XType Get_mt__Etch_Map()
	    {
		    return _mt__Etch_Map;
	    }
	
	    public XType Get_mt__Etch_Set()
	    {
		    return _mt__Etch_Set;
	    }
	
	    public XType Get_mt__Etch_Datetime()
	    {
	    	return _mt__Etch_Datetime;
	    }

        protected void addMixin(ValueFactory vf)
        {
            ValueFactory[] newMixins = new ValueFactory[mixins.Length + 1];
            System.Array.Copy(mixins,newMixins,mixins.Length);
            newMixins[mixins.Length] = vf;
            mixins = newMixins;
        }

        private ValueFactory[] mixins = { };

     

        /////////////////////
        // STRING ENCODING //
        /////////////////////

        public Encoding GetStringEncoding()
        {
            return Encoding.UTF8;
        }

        ////////////////
        // MESSAGE ID //
        ////////////////

        public long? GetMessageId(Message msg)
        {
            return (long?)(msg.Get(_mf__messageId));
        }

        public void SetMessageId(Message msg, long? msgid)
        {
            msg[_mf__messageId] = msgid;
        }

        public Field Get_mf__messageId()
        {
            return _mf__messageId;
        }

        /////////////////
        // IN REPLY TO //
        /////////////////

        public long? GetInReplyTo(Message msg)
        {
            return (long?)(msg.Get(_mf__inReplyTo));
        }

        public void SetInReplyTo(Message msg, long? msgid)
        {
            msg[_mf__inReplyTo] = msgid;
        }

        public Field Get_mf__inReplyTo()
        {
            return _mf__inReplyTo;
        }

        //////////////////////
        // VALUE CONVERSION //
        //////////////////////

       

        public Object ImportCustomValue( StructValue sv )
        {
            ImportExportHelper helper = sv.GetXType.GetImportExportHelper();

            if ( helper == null )
                return null;

            return helper.ImportValue( sv );
        }

       

        #region ValueFactory members

        public StructValue ExportCustomValue( Object value )
        {
            Type clss = value.GetType();
            XType type;
            try
            {
                type = GetCustomStructType(value.GetType());
            }
            catch (KeyNotFoundException)
            {
                type = null;
            }

            if (type == null && clss == typeof(StructValue) )
		    {
			    StructValue struct1 = (StructValue) value;
		        type = struct1.GetXType;
		    }

            if (type == null)
                //       if (value is Exception)
                if (typeof(Exception).IsAssignableFrom(clss))
                    type = _mt__Etch_RuntimeException;
                else if (typeof(IList).IsAssignableFrom(clss))
                    type = _mt__Etch_List;
                else if (typeof(IDictionary).IsAssignableFrom(clss))
                    type = _mt__Etch_Map;
                // Need to add set, but in 2.0 there is no set so just use keys of IDictionary, 
                else
                    return null;
		
		    ImportExportHelper helper = type.GetImportExportHelper();
		
		    if (helper == null)
			    return null;
		
		    return helper.ExportValue( this, value );
        }

        public XType GetCustomStructType( Type t )
        {
            XType type = class2type.Get( t );
		    if (type != null)
			    return type;
		
		    foreach (ValueFactory vf in mixins)
			    if ((type = vf.GetCustomStructType( t )) != null)
				    return type;
		
		    return null;
           
        }

        public XType GetType( int id )
        {
            XType type = types.Get(id);
            if (type != null)
                return type;

            foreach (ValueFactory vf in mixins)
                if ((type = vf.GetType(id)) != null)
                    return type;

            lock (dynamicTypes)
            {
                return dynamicTypes.Get(id);
            }
        }

        public XType GetType( string name )
        {
            XType type = types.Get(name);
            if (type != null)
                return type;

            foreach (ValueFactory vf in mixins)
                if ((type = vf.GetType(name)) != null)
                    return type;

            lock (dynamicTypes)
            {
                return dynamicTypes.Get(name);
            }
        }

        public void AddType(XType type)
        {
            lock (dynamicTypes)
            {
                dynamicTypes.Add(type);
            }
        }

        public  ICollection<XType> GetTypes()
        {
            ICollection<XType> values = types.Values();

            foreach (ValueFactory vf in mixins)
                foreach (XType x in vf.GetTypes())
                    if (!values.Contains(x))
                        values.Add(x);

            lock (dynamicTypes)
            {
                foreach (XType x in dynamicTypes.Values())
                    if (!values.Contains(x))
                        values.Add(x);
            }

            return values;
        }

        #endregion


        public void LockDynamicTypes()
        {
            dynamicTypes.Lock();
        }

        public void UnlockDynamicTypes()
        {
            // TODO dynamicTypes.Unlock();
            throw new NotSupportedException("The method or operation is not implemented.");
        }

        public Validator.Level GetLevel()
        {
            return level;
        }

        public Validator.Level SetLevel(Validator.Level level)
        {
            Validator.Level oldLevel = this.level;
            this.level = level;
            return oldLevel;
        }

        private Validator.Level level = Validator.Level.FULL;
    }
}